In [1]:
!pip install seaborn 
!pip install sklearn

In [0]:
import os 
import pandas as pd
import numpy as np 
import seaborn as sns
%matplotlib inline


from patsy import dmatrices
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import f1_score, classification_report, accuracy_score, mean_squared_error

In [0]:
# reading in Dataset
paths = 'http://bigdata.cs.byu.edu/Lectures/EEGEyeState.csv'
df = pd.read_table(paths,sep=',',encoding='latin1')

# split dependent and independent variables
y = df['eyeDetection']
X = df.iloc[:,:-1]

# create test and train.
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=.3,random_state=40)

In [6]:
# fit KNN

clf = KNeighborsClassifier()
gridsearch = GridSearchCV(clf, {"n_neighbors": [1, 3, 5, 7], "weights": ['uniform', 'distance'], 
                                'p': [1, 2, 3]}, scoring='f1')
gridsearch.fit(X_train, y_train)
`


GridSearchCV(cv=None, error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_neighbors': [1, 3, 5, 7], 'weights': ['uniform', 'distance'], 'p': [1, 2, 3]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='f1', verbose=0)

In [18]:
#best parameters
print("Best Params: {}".format(gridsearch.best_params_))
y_pred_train = gridsearch.predict(X_train)
#classification report
print("Test Classification Report:")
y_pred_test = gridsearch.predict(X_test)
print(classification_report(y_test, y_pred_test))


# Confusion Matrix
#print("Confusion Matrix")
conf_mat = pd.crosstab(y_test,y_pred_test,margins=True)
#print(conf_mat)


#sensitivity 
print('Our Sensitivity value is:', conf_mat.iloc[1,1]/conf_mat.iloc[1,2])
#specificity
print('Our Specificity is:',conf_mat.iloc[0,0]/conf_mat.iloc[0,2])
#positive predictive value 
print('Our PPV is:',conf_mat.iloc[1,1]/conf_mat.iloc[2,1])
#negative predictive value 
print('Our NPV is:',conf_mat.iloc[0,0]/conf_mat.iloc[2,0])
#accuracy
print("Test accuracy: {}".format(accuracy_score(y_test, y_pred_test)))

Best Params: {'n_neighbors': 1, 'p': 2, 'weights': 'uniform'}
Test Classification Report:
             precision    recall  f1-score   support

          0       0.97      0.97      0.97      2476
          1       0.96      0.97      0.96      2018

avg / total       0.97      0.97      0.97      4494

Our Sensitivity value is: 0.9658077304261645
Our Specificity is: 0.9693053311793215
Our PPV is: 0.9624691358024692
Our NPV is: 0.9720534629404617
Test accuracy: 0.9677347574543836


# Summary


In this model, we recognize our gridsearch cross validation algorithm to figure out what parameters are the most significant in order to optimize our models performance. From our grid search we find that...

From our accuracy results we see that we are around 97% accurate. Which is pretty good. We made sure to avoid overfitting on our data set so we accounted for that by cross validating, ie training on a part of the data and testing on a sepearate part of the data set so we don't have to worry about overfitting for this algorithm. 

Talk about specificity and sensitivity and talk about the knn model and how it works in our case. 